add launch checklist (#333)
rocicorp/replicache: launch-checklist.md

> # Replicache Launch Checklist

Reviewer: Nice. Can you create a new `doc/` folder and put it there? I'm going to be putting all the doc there.

Before you launch with Replicache in your product, it's a good idea to double-check that you have correctly covered all the small details of integration. This list can help you determine if you might have missed a detail, or deferred and then forgotten about it.

- **JS SDK**

>   - if you wish to version the schema of the mutations in `Push` and the `clientView` returned in `Pull`, version information is provided in `ReplicacheOptions.schemaVersion`

Reviewer: Grammatical nit, but this sounds informational and not like a task the dev should do. Rephrase like: "consider adding a `schemaVersion` field to `ReplicacheOptions` proactively so that you don't have to have a special case for "no schemaVersion" later.

>   - if you wish to version the schema of the mutations in `Push` and the `clientView` returned in `Pull`, version information is provided in `ReplicacheOptions.schemaVersion`

Reviewer: Please capitalize these.

>   - re-auth is handled for `Push` and `Pull` via `getPushAuth` and `getPullAuth`

Reviewer: Can your session token expire during a user session? If Replicache gets a 401 during push or pull, it will ...

- **All endpoints**

>   - the user is authenticated and the `clientID` verified to belong to that user

Reviewer: I don't understand what the task is here for the dev

- **Push endpoint**

>   - the `lastMutationID` for a client is updated transactionally along with the pushed mutations' effects

Reviewer: We need to create a *Server Requirements* doc that details exactly what is needed from server at a minimum. Like for example, I believe that CosmosDB's "consistent prefix" setting is consistent enough for Replicache's purposes.

Author: Agree we should create such a doc. In the meantime, I've included the content inline.

Author: Also PTAL at the "It is extremely important" item re consistency.

  - all mutations with `id`s less than the client's current `lastMutationID` are skipped
  - all mutations with `id`s greater than the client's current `lastMutationID+1` are ignored

>   - if a temporary error is encountered that might be resolved on retry, it halts processing mutations

Reviewer: I would phrase this more as "think about what you want the error handling policy to be" and link to a doc, either the push endpoint reference doc which discusses this or a dedicated doc on error handling.

>   - if a permanent error is encountered such that the mutation will never be appiable, ignore that mutation and increment the `lastMutationID` as if it were applied. Failing to do this can deadlock the client if a mutation will always fail.

Reviewer: sp: applicable

# Replicache Launch Checklist

Before you launch with Replicache in your product, it's a good idea to double-check that you have correctly covered all the small details of integration. This list can help you determine if you might have missed a detail, or deferred and then forgotten about one.

- **JS SDK**
  - if you wish to change the type of a mutator (eg, the number or type of its arguments) you must choose a new name; Replicache does not handle mutator versioning

>   - if you wish to version the schema of the mutations in `Push` and the `clientView` returned in `Pull`, version information is included in `ReplicacheOptions.schemaVersion`

Reviewer 2: ```suggestion
  - if you wish to version the schema of the mutations in **Push** and the **clientView** returned in **Pull**, version information is included in `ReplicacheOptions.schemaVersion`
```

Push, clientView and Pull are not methods. We do have a `pull` method but that is it.

Author: ok, i was using code style for everything that is a technical entity, believe now code style is only for actual things we have in the code.

>   - be sure that re-auth is handled for `Push` and `Pull` via `getPushAuth` and `getPullAuth`

Reviewer 2: ```suggestion
  - be sure that re-auth is handled for **Push** and **Pull** via `getPushAuth` and `getPullAuth`
```

- **All endpoints**
  - ensure the user is authenticated and the `clientID` verified to belong to that user
- **Push endpoint**
  - ensure that the `lastMutationID` for a client is updated transactionally along with the pushed mutations' effects
  - all mutations with `id`s less than the client's current `lastMutationID` must be skipped

>   - all mutations with `id`s greater than the client's current `lastMutationID+1` must be ignored

Reviewer 2: Maybe be more specific than skipped and ignored. Maybe explain why skipped/ignored.

Also, to me as a non english speaker, skipped and ignored are too subtle (I don't know the difference)

  - if a temporary error is encountered that might be resolved on retry, the endpoint must halt processing mutations and return

>   - if a permanent error is encountered such that the mutation will never be appiable, the endpoint must ignore that mutation and increment the `lastMutationID` as if it were applied. Failing to do this will deadlock the client if a mutation will always fail.

Reviewer 2: ```suggestion
  - if a permanent error is encountered such that the mutation will never be appliable, the endpoint must ignore that mutation and increment the `lastMutationID` as if it were applied. Failing to do this will deadlock the client if a mutation will always fail.
```

>   - ignore all `PushRequest`s with an unexpected `version`

Reviewer 2: ```suggestion
  - ignore all **push requests** with an unexpected `version`
```

- **Pull endpoint**
  - ensure that the `lastMutationID` returned in the response is read in the same transaction as the client view data (ie, is consistent with it)
  - if there is a problem with the `cookie` (e.g., it is unusable) return all data

>   - ignore all `PullRequest`s with an unexpected `version`

Reviewer 2: ```suggestion
  - ignore all **pull requests** with an unexpected `version`
```
